src/Nethermind/Nethermind.Consensus/Producers/PayloadAttributes.cs

```diff
@@ -118,49 +117,34 @@ protected virtual int WritePayloadIdMembers(BlockHeader parentHeader, Span<byte>
     private static PayloadAttributesValidationResult ValidateVersion(
         int apiVersion,
         int actualVersion,
-        int expectedVersion,
+        int timestampVersion,
         string methodName,
         [NotNullWhen(false)] out string? error)
     {
-        if (apiVersion >= EngineApiVersions.Cancun)
+        // version calculated from parameters should match api version
+        if (actualVersion != apiVersion)
         {
-            if (actualVersion == apiVersion && expectedVersion != apiVersion)
+            // except of Shanghai api handling Paris fork
+            if (apiVersion == EngineApiVersions.Shanghai && timestampVersion < apiVersion)
```

Comment:
What about future forks post-Shanghai?

Reply:
The rule is only for Paris <> Shanghai mismatch, I'd vote for simplification, but it requires changing Paris which is out of favor.